### DISCUSSION TITLE
Mlflow pipelines: step-specific requirements

### DISCUSSION THREAD

**Discussion by giacomov** [mlflow/mlflow#6188] [Ideas]
I was looking at the specification for the MLFlow Pipeline. The git repo of a pipeline has a top-level requirements.txt file specifying the dependencies. This is pretty limiting for a number of reasons:
* It assumes everything is written in python
* It forces all steps to share all requirements. For example, this means that a preprocessing step of a deep learning pipeline will have to install say pytorch or tensorflow even though it won't use them. It also forces all steps to use compatible dependencies, which can become intenable with time. This kind of invalidates a big feature of ML pipelines, which is the independence of steps.

It would be great if each step in a pipeline could be instead (or also) a full-blown MLFlow project, with its own conda.yaml or Dockerfile. This would solve both issues and make things also a lot more reusable across pipelines.

**Comment by jinzhang21** [comment]
Thanks for your feedback, @giacomov ! This is a very insightful feedback that we are thinking along similar directions. Are you willing to contribute to MLflow Pipelines in this area?

**Reply by ahdbilal** [reply]
@giacomov thanks for the feedback! if you did get the chance to use the experimental version of the pipeline, we would love to talk to you to get your feedback. if interested, you can use this [form](https://docs.google.com/forms/d/e/1FAIpQLSeYh6-u0qFn7AYv06Kiph09oZq2PW7Vx3Z_qVjFLE5mfMWQFA/viewform?usp=sf_link) to sign up and we will reach out to you via email.

**Comment by giacomov** [comment]
hi, yes I just filled up the form.

Sorry for the radio silence but I was enjoying some vacation :-) I am now back to work
